library/ssl_msg.c

```diff
@@ -4005,7 +4005,11 @@ static int ssl_prepare_record_content(mbedtls_ssl_context *ssl,
                  MBEDTLS_SSL_EARLY_DATA_TRY_TO_DEPROTECT_AND_DISCARD)) {
                 MBEDTLS_SSL_DEBUG_MSG(
                     3, ("EarlyData: deprotect and discard app data records."));
-                /* TODO: Add max_early_data_size check here, see issue 6347 */
+
+                ret = mbedtls_ssl_tls13_check_early_data_len(ssl, rec->data_len);
```

Comment:
Should we subtract the AEAD overhead here? And perhaps 1 for the (encrypted) content type byte as well? Even if we can't decrypt, we should be able know what the selected ciphersuite would be and deduce the AEAD overhead from that, I think.

[The RFC](https://datatracker.ietf.org/doc/html/rfc8446#page-75) says:

> Note that servers
      that reject early data due to lack of cryptographic material will
      be unable to differentiate padding from content, so clients
      SHOULD NOT depend on being able to send large quantities of
      padding in early data records.

This only mentions padding as something the server can't know the length of, so implicitly it suggest that things of which we know the length (AEAD tag, content type byte) we should subtract?

Reply:
In that case, the AEAD overhead may be extracted in `mbedtls_ssl_decrypt_buf()` if we've been far enough without error. But as we cannot decrypt and authenticate the record I am not sure we should be too smart about it here and substract one for the inner content type on top of that. What do you think?

Reply:
I still think in an ideal world we should subtract one for the inner content type (and error out if there isn't 1 byte left to subtract), but (1) I'm not feeling 100% confident about that, and (2) as discussed verbally yesterday, I think this is enough of an edge case that we don't need to worry too much about it.